src/typechecker/FStar.TypeChecker.Util.fs

```diff
@@ -93,61 +93,250 @@ let check_uvars r t =
-          then BU.print2 "(%s) Using type %s\n"
-                    (Range.string_of_range r) (Print.term_to_string t);
-          Inl t
-        | _ ->
-          Inl S.tun
-    in
-    let t =
-       match aux e with
-       | Inr c ->
-         if U.is_tot_or_gtot_comp c
-         then U.comp_result c
-         else raise_error (Errors.Fatal_UnexpectedComputationTypeForLetRec,
-                           BU.format1 "Expected a 'let rec' to be annotated with a value type; got a computation type %s"
-                                       (Print.comp_to_string c))
+  let u_subst, univ_vars = SS.univ_var_opening univ_vars in
+  let e = SS.subst u_subst e in
+  let t = SS.subst u_subst t in
+  if Env.debug env <| Options.Other "Dec"
+  then BU.print2 "extract_let_rec_annotation lbdef=%s; lbtyp=%s\n"
+                 (Print.term_to_string e)
+                 (Print.term_to_string t);
+  let env = Env.push_univ_vars env univ_vars in
+  let un_arrow t =
+    //Rather than use U.abs_formals_comp, we use un_arrow here
+    //since the former collapses adjacent Tot annotations, e.g.,
+    //    x:t -> Tot (y:t -> M)
+    // is collapsed, breaking arities
+      match (SS.compress t).n with
+      | Tm_arrow (bs, c) ->
+        Subst.open_comp bs c
+      | _ ->
+        Errors.raise_error (Errors.Fatal_LetRecArgumentMismatch, "Recursive functions must be introduced at arrow types")
                            rng
-       | Inl t -> t
+  in
+  let reconcile_let_rec_ascription_and_body_type tarr lbtyp_opt =
+      let get_decreases c =
+          U.comp_flags c |> BU.prefix_until (function DECREASES _ -> true | _ -> false)
+      in
+      match lbtyp_opt with
+      | None ->
+        let bs, c = un_arrow tarr in
+        (match get_decreases c with
+         | Some (pfx, DECREASES d, sfx) ->
+           let c = U.comp_set_flags c (pfx @ sfx) in
+           U.arrow bs c, tarr, true
+         | _ -> tarr, tarr, true)
+
+      | Some annot ->
+        let bs, c = un_arrow tarr in
+        let bs', c' = un_arrow annot in
+        if List.length bs <> List.length bs'
+        then Errors.raise_error (Errors.Fatal_LetRecArgumentMismatch, "Arity mismatch on let rec annotation")
+                                rng;
+        let move_decreases d flags flags' =
+          let d' =
+            let s = U.rename_binders bs bs' in
+            List.map (SS.subst s) d
+          in
+          let c = U.comp_set_flags c flags in
+          let tarr = U.arrow bs c in
+          let c' = U.comp_set_flags c' (DECREASES d'::flags') in
+          let tannot = U.arrow bs' c' in
+          tarr, tannot, true
+        in
+        match get_decreases c, get_decreases c' with
+        | None, _ -> tarr, annot, false
+        | Some (pfx, DECREASES d, sfx), Some (pfx', DECREASES d', sfx') ->
+          Errors.log_issue rng
+             (Warning_DeprecatedGeneric,
+              "Multiple decreases clauses on this definition; please remove the one on its declaration");
```

Comment:
The error message could also say that the one on the declaration will be ignored.